service/src/loader.rs

```diff
@@ -147,6 +147,51 @@ pub async fn build_crate(
     Ok(compilation?.cdylibs[0].path.clone())
 }
 
+pub fn clean_crate(project_path: &Path, release_mode: bool) -> anyhow::Result<Vec<String>> {
+    let (read, write) = pipe::pipe();
+    let project_path = project_path.to_owned();
+
+    tokio::task::spawn_blocking(move || {
+        let config = get_config(write).unwrap();
+        let manifest_path = project_path.join("Cargo.toml");
+        let ws = Workspace::new(&manifest_path, &config).unwrap();
+
+        let requested_profile = if release_mode {
+            InternedString::new("release")
+        } else {
+            InternedString::new("dev")
+        };
+
+        let opts = CleanOptions {
+            config: &config,
+            spec: Vec::new(),
+            targets: Vec::new(),
+            requested_profile,
+            profile_specified: true,
+            doc: false,
+        };
+
+        clean(&ws, &opts).unwrap();
+    });
+
+    let mut lines = Vec::new();
+
+    for message in Message::parse_stream(read) {
```

Comment:
I suppose this thing knows when the stream ends? That is it is not relying on the `.read(...) == Ok(0)` behaviour to determine when the task is done, otherwise this could race with the `spawn_blocking` above and, if we get here before it starts, this could return too early?

Reply:
Also, should we not put the whole thing under spawn_blocking and await completion?

Reply:
Luckily this thing knows when the stream ends, else syncing the two would have been a nightmare :sweat_smile: 

About the `spawn_blocking`, in the `build_crate` function we actually spawn block this stream. But the axum handler complains about some trait not being implemented when I do that instead. I'm not sure what that is about though :man_shrugging: